hail/src/main/scala/is/hail/expr/ir/TypeCheck.scala

```diff
@@ -5,21 +5,22 @@ import is.hail.expr.types.virtual._
 import is.hail.utils._
 
 object TypeCheck {
-  def apply(ir: IR, aggEnv: Option[Env[Type]] = None) {
-    apply(ir, new Env[Type](), aggEnv)
-  }
+  def apply(ir: BaseIR): Unit = apply(ir, Env.empty, None)
+
 
-  def apply(ir: IR, env: Env[Type], aggEnv: Option[Env[Type]]): Unit = {
+  def apply(ir: BaseIR, env: Env[Type], aggEnv: Option[Env[Type]]): Unit = {
     try {
-      _apply(ir, env, aggEnv)
+      val irCp = ir.deepCopy()
```

Comment:
I don't think I'm OK with TypeCheck copying.  We need everything to avoid sharing if we're going to use this ref equality map strategy.

Reply:
With an adjustment, I have removed this.